homeassistant/components/acmeda/sensor.py

```diff
@@ -0,0 +1,239 @@
+    # the update call is done before writing their data to the state machine.
+    #
+    # An alternative approach would be to disable automatic polling by Home
+    # Assistant and take control ourselves. This works great for polling as now
+    # we trigger from 1 time update an update to all entities. However it gets
+    # tricky from inside async_turn_on and async_turn_off.
+    #
+    # If automatic polling is enabled, Home Assistant will call the entity
+    # update method after it is done calling all the services. This means that
+    # when we update, we know all commands have been processed. If we trigger
+    # the update from inside async_turn_on, the update will not capture the
+    # changes to the second entity until the next polling update because the
+    # throttle decorator will prevent the call.
+
+    progress = None
+    cover_progress = set()
+    group_progress = set()
+
+    async def request_update(is_group, object_id):
+        """Request an update.
+
+        We will only make 1 request to the server for updating at a time. If a
+        request is in progress, we will join the request that is in progress.
+
+        This approach is possible because should_poll=True. That means that
+        Home Assistant will ask covers for updates during a polling cycle or
+        after it has called a service.
+
+        We keep track of the covers that are waiting for the request to finish.
+        When new data comes in, we'll trigger an update for all non-waiting
+        covers. This covers the case where a service is called to enable 2
+        covers but in the meanwhile some other covers has changed too.
+        """
+        nonlocal progress
+
+        progress_set = group_progress if is_group else cover_progress
+        progress_set.add(object_id)
+
+        if progress is not None:
+            return await progress
+
+        progress = asyncio.ensure_future(update_hub())
+        result = await progress
+        progress = None
+        cover_progress.clear()
+        group_progress.clear()
+        return result
+
+    async def update_hub():
+        """Update the values of the hub.
+
+        Will update covers and, if enabled, groups from the hub.
+        """
+        tasks = []
+        tasks.append(
+            async_update_items(
+                hass,
+                config_entry,
+                hub,
+                async_add_entities,
+                request_update,
+                False,
+                cur_covers,
+                cover_progress,
+            )
+        )
+
+        # tasks.append(
+        #    async_update_items(
+        #        hass,
+        #        config_entry,
+        #        hub,
+        #        async_add_entities,
+        #        request_update,
+        #        True,
+        #        cur_groups,
+        #        group_progress,
+        #    )
+        # )
+
+        await asyncio.wait(tasks)
+
+    await update_hub()
+
+
+async def async_update_items(
```

Comment:
This looks similar to the code in cover.py.  Can you break this out so its not repeated?

Reply:
Done, refactored into `helpers.py`.